metagraph/src/annotation/row_diff_builder.cpp

```diff
@@ -809,13 +811,91 @@ void convert_batch_to_row_diff(const std::string &pred_succ_fprefix,
+            total_num_labels += sources[s].num_labels();
+        }
+
+        if (total_num_labels > MAX_COLUMNS_IN_BATCH / 2) {
+            logger->warn("The number of columns to transform is large: {}."
+                         " Consider disabling disk swap (pass --disk-swap \"\").",
+                         total_num_labels);
+        }
+        if (total_num_labels > MAX_COLUMNS_IN_BATCH) {
+            logger->error("Too many columns to transform with disk swap: {} (> MAX {}).",
+                          total_num_labels, MAX_COLUMNS_IN_BATCH);
+            exit(1);
+        }
+        tmp_path = utils::create_temp_dir(swap_dir, "col");
+
+        const uint32_t chunks_open_per_thread
+                = MAX_NUM_FILES_OPEN / std::max((uint32_t)1, num_threads) / (2 + with_values);
+        if (chunks_open_per_thread < 3) {
+            logger->error("Can't merge with less than 3 chunks per thread open. "
+                          "Max num files open: {}. Current number of threads: {}.",
+                          MAX_NUM_FILES_OPEN, num_threads);
+            exit(1);
+        }
+
+        call_diffs = [&,chunks_open_per_thread](size_t s, size_t j) {
+            return [&,s,j](const std::function<void(uint64_t)> &call) {
+                std::vector<std::string> filenames;
+                // skip chunk with fwd bits which have already been counted if stage 1
+                for (uint32_t chunk = compute_row_reduction ? 1 : 0;
+                                    chunk < num_chunks[s][j]; ++chunk) {
+                    filenames.push_back(tmp_file(s, j, chunk));
+                }
+
+                const bool remove_chunks = true;
+                uint64_t r = 0;
+                elias_fano::merge_files<T>(filenames, [&](T v) {
+                    call(utils::get_first(v));
+                    if constexpr(with_values) {
+                        assert(v.second && "zero diffs must have been skipped");
+                        values[s][j][r++] = matrix::encode_diff(v.second);
+                    }
+                }, remove_chunks, chunks_open_per_thread);
+            };
+        };
+    } else {
+        logger->info("Diff-transform in memory without disk swap");
+
+        call_diffs = [&](size_t s, size_t j) {
+            return [&,s,j](const std::function<void(uint64_t)> &call) {
+                auto &v = set_rows_fwd[s][j];
+                v.insert(v.end(), set_rows_bwd[s][j].begin(), set_rows_bwd[s][j].end());
+                set_rows_bwd[s][j] = {};
+
+                std::sort(v.begin(), v.end());
+
+                uint64_t r = 0;
+                for (size_t i = 0; i < v.size(); ++i) {
+                    call(utils::get_first(v[i]));
+                    if constexpr(with_values) {
+                        assert(v[i].second && "zero diffs must have been skipped");
+                        values[s][j][r++] = matrix::encode_diff(v[i].second);
+                    }
```

Comment:
you can probably put this block in a lambda so it can be used here and above

Reply:
it would be far above, I think it's better this way